### PR TITLE
caBLE tunnel server

### PR DESCRIPTION
**Pull request by micolous**
Implements a proof-of-concept caBLE tunnel server (#259).

The basics of this seem to work OK.

**Done:**

* merge #293 
* implemented TLS support
* implemented (optional) `Origin` checks
* `backend` is functional using Chrome and Safari
* `frontend` is a basic proof-of-concept, but only supports one backend task (ie: no routing). This won't be implemented for this proof of concept
* present meaningful HTTP errors and WebSocket errors
* merge #295 
* merge #304 
* merge #305
* merge #308 

**Later:**

* integration tests

Fixes #

- [x] cargo test has been run and passes
- [x] documentation has been updated with relevant examples (if relevant)

### CODE REVIEW

File: cable-tunnel-server/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+* `common`: contains all the web server and caBLE boilerplate which is shared
+  between the `backend` and `frontend` binaries.
+
+It should be possible to run the `backend` without a `frontend` – in this case
+the routing ID will be ignored.
+
+The `frontend` should be able to handle having `backend` tasks coming and going,
+and have health checks to avoid routing it to bad `backend` tasks.
+
+This will probably need some distributed lock service to allocate the routing
```
Comment (Firstyear) [inline cable-tunnel-server/README.md:30]: Yeah, future goal. For now we don't need it though :)

Comment (micolous) [reply cable-tunnel-server/README.md:102]: Agreed :)

**Review by Firstyear — APPROVED** [review]

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
you are actually a mad lad.

**Comment by yaleman** [comment]
This is spectacular 🥳

**Pull request merged**